src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -199,44 +278,76 @@ public void PacketReset(SNIHandle handle, bool write, SNIPacket packet)
         /// <param name="async">Asynchronous connection</param>
         /// <param name="parallel">Attempt parallel connects</param>
         /// <returns>SNI handle</returns>
-        public SNIHandle CreateConnectionHandle(object callbackObject, string fullServerName, bool ignoreSniOpenTimeout, long timerExpire, out byte[] instanceName, byte[] spnBuffer, bool flushCache, bool async, bool parallel)
+        public SNIHandle CreateConnectionHandle(object callbackObject, string fullServerName, bool ignoreSniOpenTimeout, long timerExpire, out byte[] instanceName, ref byte[] spnBuffer, bool flushCache, bool async, bool parallel, bool isIntegratedSecurity)
         {
             instanceName = new byte[1];
             instanceName[0] = 0;
 
             string[] serverNameParts = fullServerName.Split(':');
 
-            if (serverNameParts.Length > 2)
+            // when no protocol specified
+            // serverNameParts.Length == 1 -> fullServerName is in hostname or ipv4 format
+            // serverNameParts.Length == 8 -> fullServerName is in ipv6 format
+            if (serverNameParts.Length == 1 || serverNameParts.Length == 8)
```

Comment:
Can IPv6 Link local address be used to connect to Sql Server? 
If yes, then this check can lead to incorrect results. For link local you can get 6 parts instead of 8.